src/librustc_typeck/check/expr.rs

```diff
@@ -871,6 +871,29 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
 
         let method = match self.lookup_method(rcvr_t, segment, span, expr, rcvr) {
             Ok(method) => {
+                // We could add a "consider `foo::<params>`" suggestion here, but I wasn't able to
+                // trigger this codepath causing `structuraly_resolved_type` to emit an error.
+
+                // We could do this only when type params are present in the method to reducte
+                // memory usage, but doing it unconditionally lets us also point at the method
+                // expression and state the resolved return value:
+                // ```
+                // error[E0282]: type annotations needed
+                //    --> $DIR/issue-65611.rs:59:20
+                //    |
+                // LL |     let x = buffer.last().unwrap().0.clone();
+                //    |             -------^^^^--
+                //    |             |      |
+                //    |             |      cannot infer type for `T`
+                //    |             this method call resolves to `std::option::Option<&T>`
+                //    |
+                //    = note: type must be known at this point
+                // ```
+                self.tables.borrow_mut().node_method_def_id_mut().insert(
```

Comment:
I *think* that this table is storing the same information that we can already get from the `type_dependent_defs` table -- we call the `write_method_call` below with `method`:

https://github.com/rust-lang/rust/blob/90b957a17c1abba979aa41234ce0993a61030e67/src/librustc_typeck/check/mod.rs#L2820-L2824

and it calls `write_resolution`

https://github.com/rust-lang/rust/blob/90b957a17c1abba979aa41234ce0993a61030e67/src/librustc_typeck/check/mod.rs#L2816-L2818

which seems to store `method.def_id` into that table.

Reply:
done